core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -313,27 +315,65 @@ private HttpResponse convertException(ServiceRequestContext ctx, HttpRequest req
+    private static Parameter createHttpComponentParameter(java.lang.reflect.Parameter parameterInfo,
                                                           ParameterType paramType, String paramValue) {
-        final Default aDefault = parameterInfo.getAnnotation(Default.class);
-        final boolean isOptionalType = parameterInfo.getType() == Optional.class;
+        assert paramType == ParameterType.PARAM ||
+               paramType == ParameterType.HEADER;
 
+        final Default aDefault = parameterInfo.getAnnotation(Default.class);
         // Set the default value to null if it was not specified.
         final String defaultValue = aDefault != null ? getSpecifiedValue(aDefault.value()).get() : null;
+
         final Class<?> type;
-        if (isOptionalType) {
+        final Class<?> wrapperType;
+
+        final Type parameterizedType = parameterInfo.getParameterizedType();
+        if (parameterizedType instanceof ParameterizedType) {
             try {
-                type = (Class<?>) ((ParameterizedType) parameterInfo.getParameterizedType())
-                        .getActualTypeArguments()[0];
-            } catch (Exception e) {
-                throw new IllegalArgumentException("Invalid optional parameter: " + parameterInfo.getName(), e);
+                type = (Class<?>) ((ParameterizedType) parameterizedType).getActualTypeArguments()[0];
+                wrapperType = validateWrapperAndElementType(paramType, parameterInfo.getType(), type);
+            } catch (Throwable cause) {
+                throw new IllegalArgumentException("Invalid optional parameter: " + parameterInfo.getName(),
+                                                   cause);
             }
         } else {
             type = parameterInfo.getType();
+            wrapperType = null;
         }
 
-        return new Parameter(paramType, !isOptionalType && aDefault == null, isOptionalType,
-                             validateAndNormalizeSupportedType(type), paramValue, defaultValue, null);
+        final boolean isRequired = wrapperType != Optional.class && aDefault == null;
+        return new Parameter(paramType, isRequired,
+                             validateAndNormalizeSupportedType(type), wrapperType,
+                             paramValue, defaultValue, null);
+    }
+
+    private static Class<?> validateWrapperAndElementType(ParameterType paramType,
+                                                          Class<?> clazz, Class<?> elementClazz) {
+        assert paramType == ParameterType.PARAM ||
+               paramType == ParameterType.HEADER;
+
+        if (clazz == Optional.class) {
+            return Optional.class;
+        }
+
+        // A list of string is supported only for HTTP headers.
+        if (paramType == ParameterType.HEADER && elementClazz == String.class) {
+            if (clazz == List.class) {
+                return ArrayList.class;
```

Comment:
Could we support:

- `Iterable` and `List` -> `ArrayList`
- `Set` -> `LinkedHashSet`

..?

Reply:
Ok, let me make a fix to support `Set<String>` as well.